lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -1515,34 +1515,45 @@ else if (value instanceof Utf8) {
 
   }
 
-  /*
+  /**
    * Called to create new array instances. Subclasses may override to use a
-   * different array implementation. By default, this returns a {@link
-   * GenericData.Array}.
+   * different array implementation. By default, this returns a
+   * {@link GenericData.Array}.
+   * 
+   * @param old    the old array instance to reuse, if possible. If the old array
+   *               is an appropriate type, it may be cleared and returned.
+   * @param size   the size of the array to create.
+   * @param schema the schema of the array elements.
```

Comment:
Note to reviewers
I have changed the spec of this method, specifically to 
- check the schema of the returned `GenericContainer` (if one is returned)

I am unsure of what other `Collection`s may be passed. Should a set be appropriate. What if the collection had different semantics? (this however is existing behaviour and unchanged by this PR)

Comparison of the schema is via `==` not `.equals` as I thought .equals would be potentially expensive, and inappropriate to have in a performance optimisation.
I appreciate that this is very opinionated, and would welcome comments from developer who know Avro better that I do

Reply:
Checking the schema of the returned container is a nice option, but I don't know if it's generally more expensive to do a full equality check as opposed to creating a new collection. I'd refrain from a change like this until we know (and then I fully agree with it).

As to other collection types: these should all implement the `Collection` interface anyway.

Reply:
Was on holiday last week, so didn't respond earlier
@opwvhk The returned schema check was just using `==` (identity check) rather than `.equals` (equality check) to avoid the cost explicitly
Can you confirm if this is OK, or should be removed as a check

My issue with he collection interface, is that reusing a `Set` will have different semantics that returning a new `List`

I guess thing sort of work OK ATM, but always wary about changing code when I know so little of the expectation of the behaviour

Reply:
@opwvhk I have removed the schema validity check as requested